cpp/serve/model.cc

```diff
@@ -284,13 +284,25 @@ class ModelImpl : public ModelObj {
                           std::move(trace_recorder));
   }
 
-  Sampler CreateSampler(int max_num_sample, Optional<EventTraceRecorder> trace_recorder) {
-    if (Sampler::SupportGPUSampler(device_)) {
-      return Sampler::CreateGPUSampler(max_num_sample, vocab_size_, &this->ft_, device_,
-                                       std::move(trace_recorder));
+  Sampler CreateSampler(int max_num_sample, std::string required_devices,
+                        Optional<EventTraceRecorder> trace_recorder) {
+    if (required_devices == "auto") {
+      if (Sampler::SupportGPUSampler(device_)) {
+        return Sampler::CreateGPUSampler(max_num_sample, vocab_size_, &this->ft_, device_,
+                                         std::move(trace_recorder));
+      } else {
+        return Sampler::CreateCPUSampler(std::move(trace_recorder));
+      }
     } else {
-      return Sampler::CreateCPUSampler(std::move(trace_recorder));
+      CHECK(required_devices == "cpu" || required_devices == "gpu");
+      if (required_devices == "cpu") {
+        return Sampler::CreateCPUSampler(std::move(trace_recorder));
+      } else if (required_devices == "gpu") {
+        return Sampler::CreateGPUSampler(max_num_sample, vocab_size_, &this->ft_, device_,
+                                         std::move(trace_recorder));
+      }
     }
+    CHECK(false) << "Invalid Sampler arguments.";
```

Comment:
See the other comment. Take in `const Array<Model>& models`, update the function body to
```c++
if (models.size() == 1 && Sampler::SupportGPUSampler(device_)) {
  // create GPU sampler and return
} else {
  // create CPU sampler and return
}
```

Reply:
Passing const Array<Model>& will cause circular reference issue as CreateSampler is ModelObject's method. So I just pass num_models.